### PR TITLE
DAOS-13169 test: Adding support for custom test yaml files.

### PR DESCRIPTION
**Pull request by phender**
In order to support removing certain test yaml file entries a new --yaml_extension launch.py argument is being added to allow users to override using the default test yaml files with any yaml files with the specified extension.

Skip-unit-tests: true
Skip-func-hw-test: true

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/launch.py [hunk 1]
```diff
@@ -465,16 +465,20 @@ class TestInfo():
...
             test_file (str): the test python file
             order (int): order in which this test is executed
+            yaml_extension (str, optional): if defined and a test yaml file exists with this
+                extension, the yaml file will be used in place of the default test yaml file.
         """
         self.name = TestName(test_file, order, 0)
         self.test_file = test_file
         self.yaml_file = ".".join([os.path.splitext(self.test_file)[0], "yaml"])
+        if yaml_extension and os.path.exists(".".join([self.yaml_file, str(yaml_extension)])):
+            self.yaml_file = ".".join([self.yaml_file, str(yaml_extension)])
```
Comment (daltonbohning) [inline src/tests/ftest/launch.py:481]: So we might have an extension like:
```
my_test.yaml.my_ext
```
Do we want to support something more generic like this?
```
my_test.my_ext.yaml
```
The main reason being that linters, etc. won't recognize "yaml.my_ext" as a yaml file

**Comment by phender** [comment]
Manual testing with `./launch.py -is -ts wolf-180 -ye custom test_always_passes test_case_1`:
```
2023/04/14 03:10:02 INFO                      list_tests: Detecting tests matching tags: avocado --paginator=off list --filter-by-tags=test_always_passes --filter-by-tags=test_case_1 ./
2023/04/14 03:10:02 DEBUG                      run_local: Running on wolf-180: avocado --paginator=off list --filter-by-tags=test_always_passes --filter-by-tags=test_case_1 ./
2023/04/14 03:10:07 DEBUG                      run_local:   wolf-180 (rc=0):
2023/04/14 03:10:07 DEBUG                      run_local:     INSTRUMENTED ./harness/basic.py:HarnessBasicTest.test_always_passes
2023/04/14 03:10:07 DEBUG                      run_local:     INSTRUMENTED ./harness/skip-list.py:TestHarnessSkipsSkipped.test_case_1
2023/04/14 03:10:07 INFO                      list_tests:   ./harness/basic.py
2023/04/14 03:10:07 INFO                      list_tests:   ./harness/skip-list.py
2023/04/14 03:10:07 INFO                             run: Modified test yaml files being created in: /tmp/tmph60qu906
2023/04/14 03:10:07 DEBUG         _add_auto_storage_yaml: Checking for auto-storage request in ./harness/basic.yaml.custom
2023/04/14 03:10:07 DEBUG         _add_auto_storage_yaml: Checking for auto-storage request in ./harness/skip-list.yaml

```

File: src/tests/ftest/launch.py (hunk 1 above)
Comment (phender) [reply src/tests/ftest/launch.py:481]: Done.

**Comment by phender** [comment]
Manual testing with extention before the `.yaml` with ``:
```
2023/04/14 03:39:28 INFO                      list_tests: Detecting tests matching tags: avocado --paginator=off list --filter-by-tags=test_always_passes --filter-by-tags=test_case_1 ./
2023/04/14 03:39:28 DEBUG                      run_local: Running on wolf-180: avocado --paginator=off list --filter-by-tags=test_always_passes --filter-by-tags=test_case_1 ./
2023/04/14 03:39:33 DEBUG                      run_local:   wolf-180 (rc=0):
2023/04/14 03:39:33 DEBUG                      run_local:     INSTRUMENTED ./harness/basic.py:HarnessBasicTest.test_always_passes
2023/04/14 03:39:33 DEBUG                      run_local:     INSTRUMENTED ./harness/skip-list.py:TestHarnessSkipsSkipped.test_case_1
2023/04/14 03:39:33 INFO                      list_tests:   ./harness/basic.py
2023/04/14 03:39:33 INFO                      list_tests:   ./harness/skip-list.py
2023/04/14 03:39:33 INFO                             run: Modified test yaml files being created in: /tmp/tmpehihjnhz
2023/04/14 03:39:33 DEBUG         _add_auto_storage_yaml: Checking for auto-storage request in ./harness/basic.custom.yaml
2023/04/14 03:39:33 DEBUG         _add_auto_storage_yaml: Checking for auto-storage request in ./harness/skip-list.yaml
...
2023/04/14 03:39:44 DEBUG                      run_local: Running on wolf-180: avocado run --ignore-missing-references --html-job-result on --tap-job-result off --filter-by-tags=test_always_passes --filter-by-tags=test_case_1 --mux-yaml ./harness/basic.custom.yaml -- ./harness/basic.py
...
2023/04/14 03:39:53 DEBUG                      run_local: Running on wolf-180: avocado run --ignore-missing-references --html-job-result on --tap-job-result off --filter-by-tags=test_always_passes --filter-by-tags=test_case_1 --mux-yaml ./harness/skip-list.yaml -- ./harness/skip-list.py
```
```
[ftest]$ ll ./harness/basic.*
-rw-r--r-- 1 root root   12 Apr 14 15:34 ./harness/basic.custom.yaml
-rw-r--r-- 1 root root 6094 Apr 12 15:52 ./harness/basic.py
-rw-r--r-- 1 root root   12 Apr 12 15:52 ./harness/basic.yaml

[ftest]$ ll ./harness/skip-list.*
-rwxr-xr-x 1 root root 4925 Apr 12 15:52 ./harness/skip-list.py
-rw-r--r-- 1 root root   12 Apr 12 15:52 ./harness/skip-list.yaml
```

**Comment by daltonbohning** [comment]
@phender Did you push changes?

File: src/tests/ftest/launch.py [hunk 2]
```diff
@@ -3164,6 +3176,10 @@ def main():
             args.logs_threshold = DEFAULT_LOGS_THRESHOLD
         args.slurm_setup = True
         args.user_create = True
+    elif args.mode == "gcp":
```
Comment (phender) [inline src/tests/ftest/launch.py:3179]: Should this be more generic?
```suggestion
    elif args.mode not in ["normal", "manual"]:
```

**Review by daltonbohning — DISMISSED** [review]

File: src/tests/ftest/launch.py (hunk 2 above)
Comment (daltonbohning) [reply src/tests/ftest/launch.py:3179]: I'm okay with that

**Review by daltonbohning — APPROVED** [review]

**Comment by phender** [comment]
One new, unrelated fault injection error: https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11955/3/Fault_Injection/new/

**Review by mjean308 — APPROVED** [review]

**Pull request merged**